### PR TITLE
Add deterministic mode test for embedding backward and use DeterministicGuard

### PR DESCRIPTION
**Pull request by albanD**
## Author Notes

Simple follow up from https://github.com/pytorch/pytorch/pull/172454 to make it clearer and keep test coverage.

## Agent Notes

The fused atomic accumulate kernel added in #172454 for `embedding_dense_backward` uses `gpuAtomicAddNoReturn`, which is non-deterministic. The kernel is correctly gated on `at::globalContext().deterministicAlgorithms()`, but the PR lacked a test verifying the deterministic fallback path and the existing tests did not properly restore the deterministic flag.

This change:
- Adds `test_embedding_backward_deterministic` that enables deterministic mode and asserts bitwise-identical gradients across 5 backward passes, using the same scenario (few unique indices, many repetitions) that would trigger the atomic kernel.
- Converts the two existing atomic-kernel tests to use `DeterministicGuard` instead of bare `torch.use_deterministic_algorithms(False)`, so the flag is properly restored on exit.
- Documents `torch.nn.Embedding` in the `use_deterministic_algorithms` docstring under operations that become deterministic when `mode=True`.

Test Plan:
```
python test/nn/test_embedding.py -k test_embedding_backward_deterministic
python test/nn/test_embedding.py -k test_embedding_backward_atomic_accumulate_kernel
```

This PR was authored with the help of an AI assistant (Claude Code).

### CODE REVIEW

File: test/nn/test_embedding.py [hunk 1]
```diff
@@ -789,41 +795,91 @@ def test_embedding_backward_atomic_accumulate_kernel_with_padding_idx(
...
+                    0,
+                    f"Expected non-zero gradient for index {idx}",
+                )
 
-        # Verify padding_idx has zero gradient
-        self.assertEqual(
-            embedding.weight.grad[padding_idx].abs().sum(),
-            0,
-            f"Expected zero gradient for padding_idx {padding_idx}",
+    @onlyOn(["cuda", "xpu"])
```
Comment (albanD) [inline test/nn/test_embedding.py:833]: Could be more but simply match the one above exactly for this PR

File: test/nn/test_embedding.py [hunk 2]
```diff
@@ -789,41 +795,91 @@ def test_embedding_backward_atomic_accumulate_kernel_with_padding_idx(
...
+
+            reference_grad = None
+            for _ in range(5):
+                out = torch.nn.functional.embedding(input=indices, weight=weight)
+                out.backward(grad_output)
+                if reference_grad is None:
+                    reference_grad = weight.grad.clone()
+                else:
+                    self.assertEqual(weight.grad, reference_grad, atol=0, rtol=0)
+                weight.grad = None
```
Comment (janeyx99) [inline test/nn/test_embedding.py:882]: i can't say i 100% get what we are testing here (it looks like we're running the same thing 5 times)--are we expecting that in nondeterministic world, one of the 5 times would be different?

**Review by janeyx99 — APPROVED** [review]
one q about the bwd test, otherwise lgtm

File: test/nn/test_embedding.py (hunk 2 above)
Comment (albanD) [reply test/nn/test_embedding.py:882]: Yes. In gradcheck we do it twice for that purpose. But here it's fast, so we do a couple more.
